File: Source/Aladdin.ISO7816.Java/src/aladdin/iso7816/ber/CRT.java
package aladdin.iso7816.ber;
import aladdin.iso7816.*;
import aladdin.iso7816.Tag;
import aladdin.asn1.*; 
import java.io.*;

///////////////////////////////////////////////////////////////////////////////
// Описание параметров алгоритма
///////////////////////////////////////////////////////////////////////////////
public class CRT extends DataObjectTemplate
{
    // конструктор закодирования
    public CRT(Tag tag, DataObject... objects)
    {    
        // сохранить переданные параметры
        super(Authority.ISO7816, tag, objects); 
        
        // проверить наличие идентификатора
        if (get(Tag.context(0x00, PC.PRIMITIVE)).length != 1) throw new IllegalArgumentException(); 
    }
    // конструктор раскодирования
    public CRT(Tag tag, TagScheme tagScheme, byte[] content) throws IOException
    {    
        // проверить корректность данных
        super(Authority.ISO7816, tag, tagScheme, content); 

        // проверить наличие идентификатора
        if (get(Tag.context(0x00, PC.PRIMITIVE)).length != 1) throw new IOException(); 
    } 
    // идентификатор алгоритма
    public final byte[] mechanismReference() 
    { 
        // идентификатор алгоритма
        return get(Tag.context(0x00, PC.PRIMITIVE))[0].content(); 
    }
    // способ использования шаблона
    public final byte usageQualifier() 
    {
        // получить способ использования шаблона
        DataObject[] objs = get(Tag.context(0x15, PC.PRIMITIVE)); 
        
        // проверить наличие способа использования
        if (objs.length == 0 || objs[0].content().length == 0) return 0; 
        
        // вернуть способ использования шаблона
        return objs[0].content()[0]; 
    }
/*    // ссылка на файл
    public final FileReference fileReference() 
    { 
        // найти объект
        BERTLV obj = get(Tag.context(0x01, PC.PRIMITIVE)); 

        // раскодировать объект
        return (obj != null) ? new FileReference(obj.content()) : null; 
    }
    // имя каталога
    public final byte[] nameDF()
    {
        // найти объект
        BERTLV obj = get(Tag.context(0x02, PC.PRIMITIVE)); 

        // раскодировать объект
        return (obj != null) ? obj.content() : null; 
    }
*/    
    ///////////////////////////////////////////////////////////////////////////
    // Описание параметров алгоритма аутентификации
    ///////////////////////////////////////////////////////////////////////////
    public static class AT extends CRT
    {
        // конструктор закодирования
        public AT(Tag tag, DataObject... objects)
        {    
            // сохранить переданные параметры
            super(tag, objects); 
            
            // проверить тип параметров
            if (!tag.equals(Tag.context(0x04, PC.CONSTRUCTED)) && 
                !tag.equals(Tag.context(0x05, PC.CONSTRUCTED)))
            {
                // при ошибке выбросить исключение
                throw new IllegalArgumentException(); 
            }
        }
        // конструктор раскодирования
        public AT(Tag tag, TagScheme tagScheme, byte[] content) throws IOException
        {    
            // проверить корректность данных
            super(tag, tagScheme, content); 

            // проверить тип параметров
            if (!tag.equals(Tag.context(0x04, PC.CONSTRUCTED)) && 
                !tag.equals(Tag.context(0x05, PC.CONSTRUCTED)))
            {
                // при ошибке выбросить исключение
                throw new IOException(); 
            }
        } 
/*        
        // идентификатор ключа
        public final byte[] secretKeyReference() 
        { 
            // найти объект
            BERTLV obj = get(Tag.context(0x03, PC.PRIMITIVE)); 

            // раскодировать объект
            return (obj != null) ? obj.content() : null; 
        }
        // идентификатор открытого ключа
        public final byte[] publicKeyReference() 
        { 
            // найти объект
            BERTLV obj = get(Tag.context(0x03, PC.PRIMITIVE)); 

            // раскодировать объект
            return (obj != null) ? obj.content() : null; 
        }
        // идентификатор личного ключа
        public final byte[] privateKeyReference() 
        { 
            // найти объект
            BERTLV obj = get(Tag.context(0x04, PC.PRIMITIVE)); 

            // раскодировать объект
            return (obj != null) ? obj.content() : null; 
        }
*/        
    }
    ///////////////////////////////////////////////////////////////////////////
    // Описание параметров алгоритма хэширования
    ///////////////////////////////////////////////////////////////////////////
    public static class HT extends CRT
    {
        // конструктор закодирования
        public HT(Tag tag, DataObject... objects)
        {    
            // сохранить переданные параметры
            super(tag, objects); 
            
            // проверить тип параметров
            if (!tag.equals(Tag.context(0x0A, PC.CONSTRUCTED)) && 
                !tag.equals(Tag.context(0x0B, PC.CONSTRUCTED)))
            {
                // при ошибке выбросить исключение
                throw new IllegalArgumentException(); 
            }
        }
        // конструктор раскодирования
        public HT(Tag tag, TagScheme tagScheme, byte[] content) throws IOException
        {    
            // проверить корректность данных
            super(tag, tagScheme, content); 

            // проверить тип параметров
            if (!tag.equals(Tag.context(0x0A, PC.CONSTRUCTED)) && 
                !tag.equals(Tag.context(0x0B, PC.CONSTRUCTED)))
            {
                // при ошибке выбросить исключение
                throw new IOException(); 
            }
        } 
/*        
        // идентификатор ключа
        public final byte[] secretKeyReference() 
        { 
            // найти объект
            BERTLV obj = get(Tag.context(0x03, PC.PRIMITIVE)); 

            // раскодировать объект
            return (obj != null) ? obj.content() : null; 
        }
*/        
    }
    ///////////////////////////////////////////////////////////////////////////
    // Описание параметров алгоритма выработки контрольной суммы
    ///////////////////////////////////////////////////////////////////////////
    public static class CCT extends CRT
    {
        // конструктор закодирования
        public CCT(Tag tag, DataObject... objects)
        {    
            // сохранить переданные параметры
            super(tag, objects); 
            
            // проверить тип параметров
            if (!tag.equals(Tag.context(0x14, PC.CONSTRUCTED)) && 
                !tag.equals(Tag.context(0x15, PC.CONSTRUCTED)))
            {
                // при ошибке выбросить исключение
                throw new IllegalArgumentException(); 
            }
        }
        // конструктор раскодирования
        public CCT(Tag tag, TagScheme tagScheme, byte[] content) throws IOException
        {    
            // проверить корректность данных
            super(tag, tagScheme, content); 

            // проверить тип параметров
            if (!tag.equals(Tag.context(0x14, PC.CONSTRUCTED)) && 
                !tag.equals(Tag.context(0x15, PC.CONSTRUCTED)))
            {
                // при ошибке выбросить исключение
                throw new IOException(); 
            }
        } 
/*        // идентификатор ключа
        public final byte[] secretKeyReference() 
        { 
            // найти объект
            BERTLV obj = get(Tag.context(0x03, PC.PRIMITIVE)); 

            // раскодировать объект
            return (obj != null) ? obj.content() : null; 
        }
*/        
    }
    ///////////////////////////////////////////////////////////////////////////
    // Описание параметров алгоритма шифрования (симметричного или асимметричного)
    ///////////////////////////////////////////////////////////////////////////
    public static class CT extends CRT
    {
        // конструктор закодирования
        public CT(Tag tag, DataObject... objects)
        {    
            // сохранить переданные параметры
            super(tag, objects); 
            
            // проверить тип параметров
            if (!tag.equals(Tag.context(0x18, PC.CONSTRUCTED)) && 
                !tag.equals(Tag.context(0x19, PC.CONSTRUCTED)))
            {
                // при ошибке выбросить исключение
                throw new IllegalArgumentException(); 
            }
        }
        // конструктор раскодирования
        public CT(Tag tag, TagScheme tagScheme, byte[] content) throws IOException
        {    
            // проверить корректность данных
            super(tag, tagScheme, content); 

            // проверить тип параметров
            if (!tag.equals(Tag.context(0x18, PC.CONSTRUCTED)) && 
                !tag.equals(Tag.context(0x19, PC.CONSTRUCTED)))
            {
                // при ошибке выбросить исключение
                throw new IOException(); 
            }
        } 
/*        
        // идентификатор ключа
        public final byte[] secretKeyReference() 
        { 
            // найти объект
            BERTLV obj = get(Tag.context(0x03, PC.PRIMITIVE)); 

            // раскодировать объект
            return (obj != null) ? obj.content() : null; 
        }
        // идентификатор открытого ключа
        public final byte[] publicKeyReference() 
        { 
            // найти объект
            BERTLV obj = get(Tag.context(0x03, PC.PRIMITIVE)); 

            // раскодировать объект
            return (obj != null) ? obj.content() : null; 
        }
        // идентификатор личного ключа
        public final byte[] privateKeyReference() 
        { 
            // найти объект
            BERTLV obj = get(Tag.context(0x04, PC.PRIMITIVE)); 

            // раскодировать объект
            return (obj != null) ? obj.content() : null; 
        }
*/
    }
    ///////////////////////////////////////////////////////////////////////////
    // Описание параметров алгоритма подписи
    ///////////////////////////////////////////////////////////////////////////
    public static class DST extends CRT
    {
        // конструктор закодирования
        public DST(Tag tag, DataObject... objects)
        {    
            // сохранить переданные параметры
            super(tag, objects); 
            
            // проверить тип параметров
            if (!tag.equals(Tag.context(0x16, PC.CONSTRUCTED)) && 
                !tag.equals(Tag.context(0x17, PC.CONSTRUCTED)))
            {
                // при ошибке выбросить исключение
                throw new IllegalArgumentException(); 
            }
        }
        // конструктор раскодирования
        public DST(Tag tag, TagScheme tagScheme, byte[] content) throws IOException
        {    
            // проверить корректность данных
            super(tag, tagScheme, content); 

            // проверить тип параметров
            if (!tag.equals(Tag.context(0x16, PC.CONSTRUCTED)) && 
                !tag.equals(Tag.context(0x17, PC.CONSTRUCTED)))
            {
                // при ошибке выбросить исключение
                throw new IOException(); 
            }
        } 
/*        // идентификатор открытого ключа
        public final byte[] publicKeyReference() 
        { 
            // найти объект
            BERTLV obj = get(Tag.context(0x03, PC.PRIMITIVE)); 

            // раскодировать объект
            return (obj != null) ? obj.content() : null; 
        }
        // идентификатор личного ключа
        public final byte[] privateKeyReference() 
        { 
            // найти объект
            BERTLV obj = get(Tag.context(0x04, PC.PRIMITIVE)); 

            // раскодировать объект
            return (obj != null) ? obj.content() : null; 
        }
*/
    }
    ///////////////////////////////////////////////////////////////////////////
    // Описание параметров алгоритма согласования ключа
    ///////////////////////////////////////////////////////////////////////////
    public static class KAT extends CRT
    {
        // конструктор закодирования
        public KAT(Tag tag, DataObject... objects)
        {    
            // сохранить переданные параметры
            super(tag, objects); 
            
            // проверить тип параметров
            if (!tag.equals(Tag.context(0x06, PC.CONSTRUCTED)) && 
                !tag.equals(Tag.context(0x07, PC.CONSTRUCTED)))
            {
                // при ошибке выбросить исключение
                throw new IllegalArgumentException(); 
            }
        }
        // конструктор раскодирования
        public KAT(Tag tag, TagScheme tagScheme, byte[] content) throws IOException
        {    
            // проверить корректность данных
            super(tag, tagScheme, content); 

            // проверить тип параметров
            if (!tag.equals(Tag.context(0x06, PC.CONSTRUCTED)) && 
                !tag.equals(Tag.context(0x07, PC.CONSTRUCTED)))
            {
                // при ошибке выбросить исключение
                throw new IOException(); 
            }
        } 
/*        // идентификатор ключа
        public final byte[] secretKeyReference() 
        { 
            // найти объект
            BERTLV obj = get(Tag.context(0x03, PC.PRIMITIVE)); 

            // раскодировать объект
            return (obj != null) ? obj.content() : null; 
        }
        // идентификатор открытого ключа
        public final byte[] publicKeyReference() 
        { 
            // найти объект
            BERTLV obj = get(Tag.context(0x03, PC.PRIMITIVE)); 

            // раскодировать объект
            return (obj != null) ? obj.content() : null; 
        }
        // идентификатор личного ключа
        public final byte[] privateKeyReference() 
        { 
            // найти объект
            BERTLV obj = get(Tag.context(0x04, PC.PRIMITIVE)); 

            // раскодировать объект
            return (obj != null) ? obj.content() : null; 
        }
*/        
    }
}

File: Source/Aladdin.ISO7816/BER/CRT.cs
using System;
using System.IO;

namespace Aladdin.ISO7816.BER
{
    ///////////////////////////////////////////////////////////////////////////////
    // Описание параметров алгоритма
    ///////////////////////////////////////////////////////////////////////////////
    public class CRT : DataObjectTemplate
    {
        // конструктор закодирования
        public CRT(Tag tag, params DataObject[] objects)

            // сохранить переданные параметры
            : base(Authority.ISO7816, tag, objects) 
        {
            // проверить наличие идентификатора
            if (this[Tag.Context(0x00, ASN1.PC.Primitive)].Length != 1) 
            {
                // при ошибке выбросить исключение
                throw new ArgumentException(); 
            }
        }
        // конструктор раскодирования
        public CRT(Tag tag, TagScheme tagScheme, byte[] content) 
            
            // проверить корректность данных
            : base(Authority.ISO7816, tag, tagScheme, content)
        {
            // проверить наличие идентификатора
            if (this[Tag.Context(0x00, ASN1.PC.Primitive)].Length != 1) 
            {
                // при ошибке выбросить исключение
                throw new InvalidDataException(); 
            }
        } 
        // идентификатор алгоритма
        public byte[] MechanismReference { get  
        { 
            // идентификатор алгоритма
            return this[Tag.Context(0x00, ASN1.PC.Primitive)][0].Content; 
        }}
        // способ использования шаблона
        public byte UsageQualifier { get  
        {
            // получить способ использования шаблона
            DataObject[] objs = this[Tag.Context(0x15, ASN1.PC.Primitive)]; 
        
            // проверить наличие способа использования
            if (objs.Length == 0 || objs[0].Content.Length == 0) return 0; 
        
            // вернуть способ использования шаблона
            return objs[0].Content[0]; 
        }}
        ///////////////////////////////////////////////////////////////////////////
        // Описание параметров алгоритма аутентификации
        ///////////////////////////////////////////////////////////////////////////
        public class AT : CRT
        {
            // конструктор закодирования
            public AT(Tag tag, params DataObject[] objects) : base(tag, objects)
            {    
                // проверить тип параметров
                if (tag != Tag.Context(0x04, ASN1.PC.Constructed) && 
                    tag != Tag.Context(0x05, ASN1.PC.Constructed))
                {
                    // при ошибке выбросить исключение
                    throw new ArgumentException(); 
                }
            }
            // конструктор раскодирования
            public AT(Tag tag, TagScheme tagScheme, byte[] content) : base(tag, tagScheme, content)
            {    
                // проверить тип параметров
                if (tag != Tag.Context(0x04, ASN1.PC.Constructed) && 
                    tag != Tag.Context(0x05, ASN1.PC.Constructed))
                {
                    // при ошибке выбросить исключение
                    throw new InvalidDataException(); 
                }
            } 
        }
        ///////////////////////////////////////////////////////////////////////////
        // Описание параметров алгоритма хэширования
        ///////////////////////////////////////////////////////////////////////////
        public class HT : CRT
        {
            // конструктор закодирования
            public HT(Tag tag, params DataObject[] objects) : base(tag, objects)
            {    
                // проверить тип параметров
                if (tag != Tag.Context(0x0A, ASN1.PC.Constructed) && 
                    tag != Tag.Context(0x0B, ASN1.PC.Constructed))
                {
                    // при ошибке выбросить исключение
                    throw new ArgumentException(); 
                }
            }
            // конструктор раскодирования
            public HT(Tag tag, TagScheme tagScheme, byte[] content) : base(tag, tagScheme, content)
            {    
                // проверить тип параметров
                if (tag != Tag.Context(0x0A, ASN1.PC.Constructed) && 
                    tag != Tag.Context(0x0B, ASN1.PC.Constructed))
                {
                    // при ошибке выбросить исключение
                    throw new InvalidDataException(); 
                }
            } 
        }
        ///////////////////////////////////////////////////////////////////////////
        // Описание параметров алгоритма выработки контрольной суммы
        ///////////////////////////////////////////////////////////////////////////
        public class CCT : CRT
        {
            // конструктор закодирования
            public CCT(Tag tag, params DataObject[] objects) : base(tag, objects)
            {    
                // проверить тип параметров
                if (tag != Tag.Context(0x14, ASN1.PC.Constructed) && 
                    tag != Tag.Context(0x15, ASN1.PC.Constructed))
                {
                    // при ошибке выбросить исключение
                    throw new ArgumentException(); 
                }
            }
            // конструктор раскодирования
            public CCT(Tag tag, TagScheme tagScheme, byte[] content) : base(tag, tagScheme, content)
            {    
                // проверить тип параметров
                if (tag != Tag.Context(0x14, ASN1.PC.Constructed) && 
                    tag != Tag.Context(0x15, ASN1.PC.Constructed))
                {
                    // при ошибке выбросить исключение
                    throw new InvalidDataException(); 
                }
            } 
        }
        ///////////////////////////////////////////////////////////////////////////
        // Описание параметров алгоритма шифрования (симметричного или асимметричного)
        ///////////////////////////////////////////////////////////////////////////
        public class CT : CRT
        {
            // конструктор закодирования
            public CT(Tag tag, params DataObject[] objects) : base(tag, objects)
            {    
                // проверить тип параметров
                if (tag != Tag.Context(0x18, ASN1.PC.Constructed) && 
                    tag != Tag.Context(0x19, ASN1.PC.Constructed))
                {
                    // при ошибке выбросить исключение
                    throw new ArgumentException(); 
                }
            }
            // конструктор раскодирования
            public CT(Tag tag, TagScheme tagScheme, byte[] content) : base(tag, tagScheme, content)
            {    
                // проверить тип параметров
                if (tag != Tag.Context(0x18, ASN1.PC.Constructed) && 
                    tag != Tag.Context(0x19, ASN1.PC.Constructed))
                {
                    // при ошибке выбросить исключение
                    throw new InvalidDataException(); 
                }
            } 
        }
        ///////////////////////////////////////////////////////////////////////////
        // Описание параметров алгоритма подписи
        ///////////////////////////////////////////////////////////////////////////
        public class DST : CRT
        {
            // конструктор закодирования
            public DST(Tag tag, params DataObject[] objects) : base(tag, objects)
            {    
                // проверить тип параметров
                if (tag != Tag.Context(0x16, ASN1.PC.Constructed) && 
                    tag != Tag.Context(0x17, ASN1.PC.Constructed))
                {
                    // при ошибке выбросить исключение
                    throw new ArgumentException(); 
                }
            }
            // конструктор раскодирования
            public DST(Tag tag, TagScheme tagScheme, byte[] content) : base(tag, tagScheme, content)
            {    
                // проверить тип параметров
                if (tag != Tag.Context(0x16, ASN1.PC.Constructed) && 
                    tag != Tag.Context(0x17, ASN1.PC.Constructed))
                {
                    // при ошибке выбросить исключение
                    throw new InvalidDataException(); 
                }
            } 
        }
        ///////////////////////////////////////////////////////////////////////////
        // Описание параметров алгоритма согласования ключа
        ///////////////////////////////////////////////////////////////////////////
        public class KAT : CRT
        {
            // конструктор закодирования
            public KAT(Tag tag, params DataObject[] objects) : base(tag, objects)
            {    
                // проверить тип параметров
                if (tag != Tag.Context(0x06, ASN1.PC.Constructed) && 
                    tag != Tag.Context(0x07, ASN1.PC.Constructed))
                {
                    // при ошибке выбросить исключение
                    throw new ArgumentException(); 
                }
            }
            // конструктор раскодирования
            public KAT(Tag tag, TagScheme tagScheme, byte[] content) : base(tag, tagScheme, content)
            {    
                // проверить тип параметров
                if (tag != Tag.Context(0x06, ASN1.PC.Constructed) && 
                    tag != Tag.Context(0x07, ASN1.PC.Constructed))
                {
                    // при ошибке выбросить исключение
                    throw new InvalidDataException(); 
                }
            } 
        }
    }
}
